### PR TITLE
[RUN-3112] Fix sudo hang and exit-code parse crash in SSHJ executor

### PR DESCRIPTION
**Pull request by ncofreortiz-hub**
## Summary

Fixes the sudo-over-PTY flow introduced in [PR #66](https://github.com/rundeck-plugins/sshj-plugin/pull/66) (Rundeck 5.8.0, RPL-40), which customers reported as either hanging forever or failing with `Failed: Unknown: For input string: "[?2004l0"`.

Three separate defects in `SudoCommand.java`, all still present on `main`:

1. **Bogus timeout unit** — `.withTimeout(30000, TimeUnit.SECONDS)` is ~8.3 hours, not 30 seconds. Any failed prompt match therefore looked like an infinite hang instead of failing fast. Fixed to `TimeUnit.MILLISECONDS`.
2. **Fragile prompt regex** — `PROMPT_PATTERN = "~.*\\$"` assumes every shell prompt contains a literal `~` and ends in `$`. Root shells (`#`) and custom `PS1` values never match, so the initial `expect()` blocks. Broadened to `.*[#$]\s*`.
3. **Unsanitized exit-code parsing** — `expectit`'s `removeNonPrintable()` filter strips only the leading ESC byte of an ANSI control sequence (e.g. a bracketed-paste-mode toggle), leaving the visible remainder (`[?2004l`) glued to the real output. That string was fed directly into `Integer.parseInt()` with no validation, producing the reported `NumberFormatException`. Added `sanitizeExitCode()`/`isValidExitCode()` to strip escape remnants and raise a clear `IOException` instead of an opaque parse crash if the response is still not a plain integer.

## Test plan

- [x] `./gradlew compileJava compileTestGroovy` — compiles clean
- [x] `./gradlew test --tests SudoCommandSpec --tests SSHJExecSpec` — 14/14 pass (new `SudoCommandSpec` covers all three fixes + regex/sanitize edge cases; existing `SSHJExecSpec` regression still green)
- [ ] Manual end-to-end verification against a real RHEL9/Ubuntu node with password auth (not run in this environment — no live SSH target available)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/main/java/com/plugin/sshjplugin/sudo/SudoCommand.java [hunk 1]
```diff
@@ -24,7 +25,32 @@ public class SudoCommand {
     private String sudoPassword;
     private PluginLogger logger;
 
-    private final String PROMPT_PATTERN = "~.*\\$";
+    // Matches a shell prompt ending in dollar sign (regular user) or hash (root),
+    // optionally followed by trailing whitespace. The previous pattern "~.*\\$"
+    // assumed every prompt contains a literal '~' (e.g. "user@host:~$"), which is
+    // not true for custom PS1 values or root shells, causing the initial expect()
+    // to never match and block until the (misconfigured) timeout elapsed.
+    static final String PROMPT_PATTERN = ".*[#$]\\s*";
```
[the remark this answers is not in the document]
Comment (ncofreortiz-hub) [reply src/main/java/com/plugin/sshjplugin/sudo/SudoCommand.java]: Fixed in 3720d42 - anchored PROMPT_PATTERN with `$` so it requires the match to reach the end of the buffered output, not just contain a `$`/`#` anywhere. Added a regression test using `Matcher.find()` (matching expectit's actual matching semantics) proving "export PATH=$PATH:..." no longer matches.

File: src/main/java/com/plugin/sshjplugin/sudo/SudoCommand.java [hunk 2]
```diff
@@ -24,7 +25,32 @@ public class SudoCommand {
...
+    static String sanitizeExitCode(String raw) {
+        return ANSI_ESCAPE_REMNANT.matcher(raw).replaceAll("").trim();
+    }
+
+    /**
+     * True if the sanitized response is a plain integer, as a shell exit code must be.
+     */
+    static boolean isValidExitCode(String sanitized) {
+        return sanitized.matches("-?\\d+");
+    }
```
[the remark this answers is not in the document]
Comment (ncofreortiz-hub) [reply src/main/java/com/plugin/sshjplugin/sudo/SudoCommand.java:72]: Fixed in 3720d42 - isValidExitCode now calls Integer.parseInt() directly in a try/catch instead of a digit regex, so it rejects anything that would overflow int (e.g. a 19-digit string), matching exactly what the caller does. Added a test case for this.

File: src/main/java/com/plugin/sshjplugin/sudo/SudoCommand.java [hunk 3]
```diff
@@ -74,7 +112,10 @@ public String runSudoCommand(String command) throws IOException {
                 //.withEchoInput(new SSHJAppendable(pluginLogger,2))
                 .withInputFilters(removeColors(), removeNonPrintable())
                 .withExceptionOnFailure()
-                .withTimeout(30000, TimeUnit.SECONDS)
+                // Was "30000, TimeUnit.SECONDS" (~8.3 hours) - a unit typo that made
+                // the sudo flow appear to hang indefinitely instead of failing fast
+                // when a prompt/pattern never matched.
+                .withTimeout(30000, TimeUnit.MILLISECONDS)
```
Comment (ronaveva) [inline src/main/java/com/plugin/sshjplugin/sudo/SudoCommand.java]: Is this TO enough for any kind of command to be triggered? @ncofreortiz-hub

File: src/main/java/com/plugin/sshjplugin/sudo/SudoCommand.java [hunk 4]
```diff
@@ -90,17 +152,34 @@ public String runSudoCommand(String command) throws IOException {
...
 
-        expect.expect(regexp(PROMPT_PATTERN));
+        // Waiting for the shell prompt to return here means waiting for the sudo'd
+        // command itself to finish, which can take far longer than a prompt
+        // exchange - so this uses the user-configurable command timeout (same
+        // ssh-command-timeout knob the non-sudo path honors) instead of the short
+        // prompt-detection default above.
+        Expect commandWait = commandTimeoutMs > 0
+                ? expect.withTimeout(commandTimeoutMs, TimeUnit.MILLISECONDS)
+                : expect.withInfiniteTimeout();
```
Comment (ronaveva) [inline src/main/java/com/plugin/sshjplugin/sudo/SudoCommand.java:162]:
```suggestion
                : expect.withTimeout(Long.MAX_VALUE, TimeUnit.DAYS);
```

since withInfiniteTimeout is deprecated, we could use something like this

**Review by ronaveva — APPROVED** [review]
LGTM, make sure to solve the pending observation before merging
